frontend/readcollection.lua

```diff
@@ -2,149 +2,220 @@ local DataStorage = require("datastorage")
+    self.coll[collection_name][item.file] = item
+    self:write(collection_name)
+end
+
+function ReadCollection:addItems(files, collection_name) -- files = { filepath = true, }
+    collection_name = collection_name or self.default_collection_name
+    local coll = self.coll[collection_name]
+    local max_order = self:getCollectionMaxOrder(collection_name)
+    for file in pairs(files) do
+        max_order = max_order + 1
+        local item = buildEntry(file, max_order)
+        coll[item.file] = item
+    end
+    self:write(collection_name)
+end
+
+function ReadCollection:removeItem(file, collection_name, no_write)
+    local coll_name, file_name = self:getFileCollectionName(file, collection_name)
+    if coll_name then
+        self.coll[coll_name][file_name] = nil
+        if not no_write then
+            self:write(coll_name)
+        end
+        return true
+    end
+end
+
+function ReadCollection:removeItems(files) -- files = { filepath = true, }
+    local do_write
+    for file in pairs(files) do
+        if self:removeItem(file, nil, true) then
+            do_write = true
+        end
+    end
+    if do_write then
+        self:write()
+    end
+end
+
+function ReadCollection:removeItemsByPath(path)
+    local do_write
+    for coll_name, coll in pairs(self.coll) do
+        for file_name in pairs(coll) do
+            if util.stringStartsWith(file_name, path) then
+                self.coll[coll_name][file_name] = nil
+                do_write = true
+            end
+        end
+    end
+    if do_write then
+        self:write()
+    end
 end
 
-function ReadCollection:checkItemExist(item, collection_name)
-    local coll = self:read(collection_name)
-    for _, v in pairs(coll) do
-        if v.file == item then
-            return true
+function ReadCollection:_updateItem(coll_name, file_name, new_filepath, new_path)
+    local coll = self.coll[coll_name]
+    local item_old = coll[file_name]
+    local order, mandatory = item_old.order, item_old.mandatory
+    new_filepath = new_filepath or new_path .. "/" .. item_old.text
+    coll[file_name] = nil
+    local item = buildEntry(new_filepath, order, mandatory) -- no lfs call
+    coll[item.file] = item
+end
+
+function ReadCollection:updateItem(file, new_filepath)
+    local coll_name, file_name = self:getFileCollectionName(file)
+    if coll_name then
+        self:_updateItem(coll_name, file_name, new_filepath)
+        self:write(coll_name)
+    end
+end
+
+function ReadCollection:updateItems(files, new_path) -- files = { filepath = true, }
+    local do_write
+    for file in pairs(files) do
+        local coll_name, file_name = self:getFileCollectionName(file)
+        if coll_name then
+            self:_updateItem(coll_name, file_name, nil, new_path)
+            do_write = true
         end
     end
+    if do_write then
+        self:write()
+    end
 end
 
+function ReadCollection:updateItemsByPath(path, new_path)
+    path = "^" .. path .. "/"
```

Comment:
(Reminds me of another recent comment :)
Possible risk if using :gsub() and path contains some pattern, might be safer to use string comparison like util.stringStartsWith() ?

Reply:
Yes, I remember https://github.com/koreader/koreader/pull/11090#discussion_r1387590240 about `string.match`.
But here we need to do actual `gsub`, to replace old path with the new one.
One more place is
https://github.com/koreader/koreader/blob/dce12de09d4c4030954ce00e72f3986c89a2a932/frontend/readhistory.lua#L194-L198